### PR TITLE
Add `plist` Parser

### PR DESCRIPTION
**Pull request by toshi0383**
Related: https://github.com/SwiftGen/SwiftGen/pull/288

### CODE REVIEW

**Comment by toshi0383** [comment]
Separated commit for source code fix and `pod install` diff.
Please let me know if I did anything wrong.
I will update the submodule url after https://github.com/SwiftGen/templates/pull/34 is merged.

File: Sources/Parsers/PlistParser.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  var keys: [String] = []
+
+  public init() {}
+
+  public func parse(at path: Path) {
+    guard let keys = loadPlistKeys(at: path) else { return }
+    self.keys = keys
+  }
+
+  fileprivate func loadPlistKeys(at path: Path) -> [String]? {
```
Comment (djbe) [inline Sources/Parsers/PlistParser.swift]: Why not make this private? Also: we're in a plist parser, no need to repeat this --> `loadKeys`

File: Sources/Parsers/PlistParser.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+//  Created by Toshihiro Suzuki on 4/12/17.
+//  Copyright © 2017 SwiftGen. All rights reserved.
+//
+
+import Foundation
+import PathKit
+
+public final class PlistParser {
+
+  var keys: [String] = []
```
Comment (djbe) [inline Sources/Parsers/PlistParser.swift]: Might be more interesting to also load & store type information for each key. Another thing you might want to do: nice names for keys (like the Xcode plist editor).

Also: try to support multiple files, like the other parsers.

File: Sources/Stencil/PlistContext.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+//  SwiftGenKit
+//
+//  Created by Toshihiro Suzuki on 4/12/17.
+//  Copyright © 2017 SwiftGen. All rights reserved.
+//
+
+import Foundation
+extension PlistParser {
+  public func stencilContext() -> [String: Any] {
+    return ["plistKeys": keys]
```
Comment (djbe) [inline Sources/Stencil/PlistContext.swift]: See comment above: try to support multiple files, and again: we're in a plist parser, don't repeat that information.

File: Tests/SwiftGenKitTests/PlistTests.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  func testEmpty() {
+    let parser = PlistParser()
+
+    let result = parser.stencilContext()
+    let expected = [
+      "plistKeys": []
+    ]
+
+    XCTDiffContexts(result, expected)
+  }
```
Comment (djbe) [inline Tests/SwiftGenKitTests/PlistTests.swift]: newline please.

**Review by djbe — CHANGES_REQUESTED** [review]

File: Tests/SwiftGenKitTests/PlistTests.swift (hunk 4 above)
Comment (toshi0383) [reply Tests/SwiftGenKitTests/PlistTests.swift]: Sure, I'll fix this.

File: Sources/Parsers/PlistParser.swift (hunk 1 above)
Comment (toshi0383) [reply Sources/Parsers/PlistParser.swift]: Sure, I'll fix this.

File: Sources/Parsers/PlistParser.swift (hunk 2 above)
Comment (toshi0383) [reply Sources/Parsers/PlistParser.swift]: Thanks for the review!
I completely agree with you, but I think I should do the fix in different PR.
This PR is enough for introducing `plist` command, at minimum. We can always improve in next release, supporting type inference and multiple files.

**Comment by djbe** [comment]
To address your point of wanting to merge a small PR now, and add more features in PRs later:

I'd usually completely agree with you, but we also have to think long term. Anytime we change released functionality, we have to do major version bumps, and break things for existing users. You could say the changes'd be minimal, but even small changes matter. I'd prefer if we work a bit longer on this feature, before actually merging (and releasing) it.

Anyway, the current timeline for SwiftGen is a small bugfix release (4.2.1), hopefully soon. We won't be adding any new major features for that release. After that we're planning on a bigger 5.0 release, where this feature could be merged into. The timeline for 5.0 is as you can guess, still a while away, so don't be in a hurry to get this merged, take your time.

TL;DR: You have time to add functionality to this PR, such as type information and multiple files.

**Comment by toshi0383** [comment]
Thank you for sharing the schedule. Understood.
I'll work on adding more functionality and clean ups!👍

**Comment by toshi0383** [comment]
As mentioned in SwiftGen's PR, added type inference support.

Also added arbitrary Bundle support, instead of multiple files support.
IMO, parsing multiple Info.plist for a single Bundle has no points, because Info.plist is officially accessible only via `bundle.infoDictionary`.
So supporting arbitrary Bundle makes more sense to me.
See https://github.com/SwiftGen/SwiftGen/pull/288#issuecomment-295517320 for use case.

**Comment by djbe** [comment]
Implemented by https://github.com/SwiftGen/SwiftGen/pull/379.
